pysoa/client/client.py

```diff
@@ -169,6 +180,45 @@ def get_all_responses(self, receive_timeout_in_seconds=None):
         finally:
             self.metrics.commit()
 
+    def get_response_for_request(self, expected_request_id, receive_timeout_in_seconds=None):
+        """
+        Receive the response for a particular request.
+
+        :param expected_request_id: ID of a request previously made via send_request.
+        :type expected_request_id: int
+        :param receive_timeout_in_seconds: How long to block without receiving a message before raising
+                                           `MessageReceiveTimeout` (defaults to five seconds unless the settings are
+                                           otherwise).
+        :type receive_timeout_in_seconds: int
+
+        :return: The expected JobResponse object.
+        :rtype: JobResponse
+
+        :raise: ConnectionError, MessageReceiveError, MessageReceiveTimeout, InvalidMessage, StopIteration
+        """
+        unexpected_responses = OrderedDict()
+
+        try:
+            start_time = time.time()
+            if receive_timeout_in_seconds is None:
+                end_time = six.sys.maxsize
+            else:
+                end_time = start_time + receive_timeout_in_seconds
+
+            while start_time < end_time:
+                receive_timeout_in_seconds = int(end_time - start_time) if receive_timeout_in_seconds is not None else None
+                for request_id, response in self.get_all_responses(receive_timeout_in_seconds):
+                    if request_id != expected_request_id:
+                        unexpected_responses[request_id] = response
+                    else:
+                        return response
+                start_time = time.time()
+        finally:
+            self.unexpected_responses = unexpected_responses
```

Comment:
Not sure if I understand what it does correctly. 
Let's say you you send jobs `1`, `2` & `3`. The first two are slow, the last one is fast. 
So let's say servers responds in order `3`, `2`, `1`.

Client waiting for job `1` will get responses for jobs `3` and `2`, put it on local `unexpected_responses` and not share to `self.unexpected_responses` until either response for job `1` or timeout comes. Why isn't line 212 directly `self.unexpected_responses`?

Reply:
We're building a list of unexpected responses we see (as opposed to the expected response that we want to retrieve). At the very end of this function, we set the unexpected responses we saw to the list of responses we should return first via `_get_responses` (vs. getting new ones from the transport).